jupyterlab_bokeh_server/server.py

```diff
@@ -141,12 +144,177 @@ def cb():
+            item_dict["memory-"+str(i)] = []
+
+        source = ColumnDataSource( item_dict )
+        def _get_color( ind ):
+            color_list = [ "blue", "red", "green", "black", "brown", "cyan", \
+                           "orange", "pink", "purple", "gold" ]
+            return color_list[ ind % len(color_list) ]
+
+        memory_fig = figure(
+            title="Memory Utilization",
+            sizing_mode="stretch_both",
+            x_axis_type="datetime",
+            y_range=[0, gpu_mem_max],
+            x_range=x_range,
+            tools=tools,
+        )
+        for i in range(ngpus):
+            memory_fig.line(source=source, x="time", y="memory-"+str(i), color=_get_color(i), legend="GPU-"+str(i))
+        memory_fig.yaxis.formatter = NumeralTickFormatter(format="0.0b")
+        memory_fig.legend.location = "top_left"
+
+        gpu_fig = figure(
+            title="GPU Utilization [%]",
+            sizing_mode="stretch_both",
+            x_axis_type="datetime",
+            y_range=[0, 100],
+            x_range=x_range,
+            tools=tools,
+        )
+        for i in range(ngpus):
+            gpu_fig.line(source=source, x="time", y="gpu-"+str(i), color=_get_color(i), legend="GPU-"+str(i))
+        gpu_fig.legend.location = "top_left"
+
+        tot_fig = figure(
+            title="Total Utilization [%]",
+            sizing_mode="stretch_both",
+            x_axis_type="datetime",
+            y_range=[0, 100],
+            x_range=x_range,
+            tools=tools,
+        )
+        tot_fig.line(source=source, x="time", y="gpu-total", color="blue", legend="Total-GPU")
+        tot_fig.line(source=source, x="time", y="memory-total", color="red", legend="Total-Memory")
+        tot_fig.legend.location = "top_left"
+
+        doc.title = "Resource Timeline"
+        doc.add_root(
+            column(gpu_fig, memory_fig, tot_fig, sizing_mode="stretch_both")
+        )
 
+        last_time = time.time()
+
+        def cb():
+            nonlocal last_time
+            now = time.time()
+            src_dict = {"time": [now * 1000]}
+            gpu_tot = 0
+            mem_tot = 0
+            for i in range(ngpus):
+                gpu = nvmlDeviceGetUtilizationRates( gpu_handles[i] ).gpu
+                mem = nvmlDeviceGetMemoryInfo( gpu_handles[i] ).used
+                gpu_tot += gpu
+                mem_tot += mem / (1024*1024)
+                src_dict["gpu-"+str(i)] = [gpu]
+                src_dict["memory-"+str(i)] = [mem]
+            src_dict["gpu-total"] = [gpu_tot / ngpus]
+            src_dict["memory-total"] = [(mem_tot/gpu_mem_sum)*100]
+
+            source.stream(
+                src_dict,
+                1000,
+            )
+
+            last_time = now
+
+        doc.add_periodic_callback(cb, 200)
+    routes = {
+        "/cpu": cpu,
+        "/resources": resource_timeline,
+        "/gpu": gpu,
+        "/gpu_resources": gpu_resource_timeline,
```

Comment:
We should probably make the label a bit nicer for multi-word endpoints so that it shows up as "GPU RESOURCES" rather than "GPU_RESOURCES"